store/postgres/src/chain_store.rs

```diff
@@ -1813,6 +1841,164 @@ impl ChainStoreTrait for ChainStore {
     }
 }
 
+mod recent_blocks_cache {
+    use super::*;
+    use std::collections::BTreeMap;
+
+    struct CachedBlock {
+        ptr: BlockPtr,
+        data: Option<json::Value>,
+        parent_hash: BlockHash,
+    }
+
+    struct Inner {
+        // Note: At all times, only a continuous range of blocks will be in the
+        // cache, i.e. without gaps. This constraint makes it possible to easily
+        // verify a direct line of ancestry between two blocks via parent hashes.
+        blocks: BTreeMap<BlockNumber, CachedBlock>,
+        // We only store these many blocks.
+        capacity: usize,
+    }
+
+    impl Inner {
+        fn get_block(
+            &self,
+            child: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(&BlockPtr, Option<&json::Value>)> {
+            // Before we can go find the ancestor, we need to make sure that
+            // we're looking for the ancestor of the right block, i.e. check if
+            // the hash (and number) of the child matches.
+            let child_is_cached = &self.blocks.get(&child.number)?.ptr == child;
+
+            if child_is_cached {
+                let ancestor_block_number = child.number - offset;
+                let block = self.blocks.get(&ancestor_block_number)?;
+                Some((&block.ptr, block.data.as_ref()))
+            } else {
+                None
+            }
+        }
+
+        fn chain_head(&self) -> Option<&BlockPtr> {
+            self.blocks.last_key_value().map(|b| &b.1.ptr)
+        }
+
+        fn earliest_block(&self) -> Option<&CachedBlock> {
+            self.blocks.first_key_value().map(|b| b.1)
+        }
+
+        fn evict_if_necessary(&mut self) {
+            while self.blocks.len() > self.capacity && !self.blocks.is_empty() {
+                self.blocks.pop_first();
+            }
+        }
+
+        fn insert_block(
+            &mut self,
+            ptr: BlockPtr,
+            data: Option<json::Value>,
+            parent_hash: BlockHash,
+        ) {
+            fn is_parent_of(parent: &BlockPtr, child: &CachedBlock) -> bool {
+                child.ptr.number == parent.number + 1 && child.parent_hash == parent.hash
```

Comment:
`+ 1` and `- 1` on block number is not correct, NEAR for example don't have strictly continuous block number, a valid chain in NEAR can be `#14 (Hash ABC1, Parent XX) -> #17 (Hash EBD2, Parent ABC1)` is valid.

I imagine this code will be hit by NEAR, so it's important to have in mind. I did not check the rest of this PR, but it review of `+ 1` nor `- 1` logic on block number should be carefully reviewed.

Reply:
Good point. I think the checks by number are redundant anyway since the hash/parent hash chain should be enough to establish these relationships.

Reply:
Good catch! Fixed with https://github.com/graphprotocol/graph-node/pull/4215/commits/6d46bacb80dc312f9cfb32c49b581b3ce33c571c. By relying entirely on parent hashes without any block number comparison, I think we should be good.